Fix flaky E2E tests in Contact Management and Email Management suites

matbcvo opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🟢 <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | 🔴
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #14174 <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

This PR fixes two flaky E2E tests:

1. **ContactManagementCest: Batch set 'Do Not Contact'**
2. **EmailManagementCest: Try to batch change email category**

![image](https://github.com/user-attachments/assets/7f75a08b-7409-4302-ab30-35979d10676a)

**Changes Made**
1. **Contact Management: Batch Set 'Do Not Contact'**
   - **Issue:** The test intermittently failed to verify that both contacts were marked as "Do Not Contact."
   - **Fixes:**
     - Added waitForElementVisible to verify elements are visible.
     - The selector for the second contact's "Do Not Contact" indicator was incorrectly pointing to the first contact's element.
     - Verified contact selection by asserting checkboxes are checked before proceeding.
     - Added wait times for any UI transitions or backend processing between steps.

2. **Email Management: Batch Change Email Category**
   - **Issue:** The test intermittently failed to verify that all emails were assigned the new category.
   - **Fixes:**
     - Verified all contacts selection by asserting checkbox is checked before proceeding.
     - Added wait times for any UI transitions or backend processing between steps.

**Testing**
Both tests were run multiple times in the CI environment, confirming consistent results.

---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Run the tests in this PR repeatedly to check if either of the two tests fails.

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

matbcvo committed:
attempt to fix flaky e2e test

PatrickJenkner reviewed (approved):
I ran the tests a couple of times and there aren't any failures, so looks good 👍 .
The code changes make sense as well. Having to use a wait with fixed length is always unfortunate, but it does the trick

escopecz reviewed (commented):
Thanks for looking into this! It was driving me crazy!

Adding the waits will make the tests less flaky but also way slower. Isn't there some smarter way how to achieve this? Why the `waitForElementVisible` methods aren't good enough?

matbcvo committed:
refactor, remove waits, add waitForElement*, notification check

matbcvo committed:
replace wait with waitForElementVisible after page reload

matbcvo commented:
@escopecz You're not alone, it was driving me crazy too while debugging those flaky E2E tests! I removed all the `wait` method I previously added and replaced them with `waitForElement*` methods. Some steps still failed, and the issue was with `reloadPage`. Previously, I was using `wait` before reloading the page, but I figured out that adding a check for the notification that pops up after a change is made allowed me to avoid using `wait`.

To verify the fix, I used a GitHub workflow that ran the E2E tests 10 times consecutively, and none of them failed. This confirms that the changes are sufficient to stabilize the E2E tests.

I suggest implementing a similar GitHub workflow to automatically detect flaky E2E tests whenever someone implements or modifies them. The workflow can be configured to trigger only when files in the `tests` folder are modified. Let me know if you think this is worth implementing, and I can create a new PR to add this workflow.

![image](https://github.com/user-attachments/assets/c049c659-2ee5-4256-b4e1-bea696d80df2)
<details>
  <summary>Workflow file for running E2E tests X times</summary>

```
name: e2e tests

on:
  push:
    paths:
      - 'tests/**'
  pull_request:
    paths:
      - 'tests/**'
  workflow_dispatch:

permissions:
  contents: read # to fetch code (actions/checkout)

jobs:
  
  e2e-tests:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        iteration: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] # run 10 times
      fail-fast: true

    steps:
    - name: Checkout
      uses: actions/checkout@v4
      with:
        fetch-depth: 0

    - name: Setup DDEV
      uses: ddev/github-action-setup-ddev@v1

    - name: Set up environment variables for test mode
      run: echo -e "APP_ENV=test\nAPP_DEBUG=1" > .env.local

    - name: Run E2E tests
      run: ddev exec bin/codecept run acceptance

    - name: Upload test artifacts
      if: failure()
      uses: actions/upload-artifact@v4
      with:
        name: e2e-test-screenshots
        path: tests/_output
        if-no-files-found: ignore
```
</details>

matbcvo committed:
remove unnecessary seeElement line

matbcvo committed:
refactor

escopecz reviewed (approved):
@matbcvo you are a life saver! Thank you so much!

Yes, I believe such new workflow would be helpful so we avoid flaky tests in the future.
